### ISSUE TITLE
Build HPX with Apple's clang

### ISSUE THREAD

**Issue by fmilicchio** [STEllAR-GROUP/hpx#2026]
Dear all,

I've searched through the issues, but I cannot find a way to build HPX with Apple's provided `clang`. These are the versions of Xcode and its compiler:

```
Apple LLVM version 7.0.2 (clang-700.1.81)
Target: x86_64-apple-darwin15.3.0
Thread model: posix

Xcode Version 7.2.1 (7C1002)
```

The error is about `thread_local`, it seems it is still not supported even if the correct flags are provided to the compilation process:

```
% VERBOSE=1 make
[...]
[  0%] Building CXX object src/CMakeFiles/hpx.dir/exception.cpp.o
cd /Users/sensei/Downloads/hpx_0.9.11/tmp/src && /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/c++   -DHPX_COMPONENT_NAME=hpx -DHPX_COROUTINE_EXPORTS -DHPX_ENABLE_ASSERT_HANDLER -DHPX_EXPORTS -DHPX_LIMIT=5 -D_GNU_SOURCE -Dhpx_EXPORTS -I/Users/sensei/Downloads/hpx_0.9.11 -I/Users/sensei/Downloads/hpx_0.9.11/tmp -I/usr/local/include -I/Users/sensei/Downloads/hpx_0.9.11/external/cache -I/usr/local/Cellar/hwloc/1.11.2/include  -std=c++14 -ftemplate-depth=256 -Wall -Wextra -Wno-unused-local-typedefs -Wno-strict-aliasing -Wno-sign-promo -Wno-attributes -Wno-cast-align -Wno-unused-parameter -Wno-ignored-qualifiers -Wformat=2 -Wno-format-nonliteral -Winit-self -Wcast-qual -Werror=parentheses -Werror=reorder -Werror=return-type -Werror=sequence-point -Werror=format -Werror=missing-braces -Werror=sign-compare -fdiagnostics-show-option -Werror=vla -Werror=return-type -Wno-cast-align -O3 -DNDEBUG -DHPX_DISABLE_ASSERTS -DBOOST_DISABLE_ASSERTS -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.11.sdk -fPIC   -o CMakeFiles/hpx.dir/exception.cpp.o -c /Users/sensei/Downloads/hpx_0.9.11/src/exception.cpp
In file included from /Users/sensei/Downloads/hpx_0.9.11/src/exception.cpp:7:
In file included from /Users/sensei/Downloads/hpx_0.9.11/hpx/hpx_fwd.hpp:57:
In file included from /Users/sensei/Downloads/hpx_0.9.11/hpx/runtime/threads_fwd.hpp:14:
In file included from /Users/sensei/Downloads/hpx_0.9.11/hpx/runtime/threads/thread_data_fwd.hpp:14:
In file included from /Users/sensei/Downloads/hpx_0.9.11/hpx/util/coroutine/detail/default_context_impl.hpp:109:
In file included from /Users/sensei/Downloads/hpx_0.9.11/hpx/util/coroutine/detail/context_generic_context.hpp:34:
In file included from /usr/local/include/boost/context/all.hpp:13:
In file included from /usr/local/include/boost/context/execution_context.hpp:10:
/usr/local/include/boost/context/execution_context.ipp:61:5: error: thread-local storage is not supported for the
      current target
    thread_local static ptr_t   current_rec;
    ^
1 error generated.
```

I have tried adding `-stdlib=libc++` to `CMAKE_CXX_FLAGS`, but it won't make it. 

Has anyone succeeded in building HPX with Apple-provided compiler?

Thanks!

**Comment by sithhell** [comment]
This seems to be a problem with boost. Which Boost version are you using? We are testing AppleClang 6.1.0.6020053 with boost version 1.56.

**Comment by sithhell** [comment]
In addition, would you mind checking the latest head of master instead of the release?

**Comment by hkaiser** [comment]
Discussed yesterday on IRC:

[13 Mar 16 15:33] * zao * hkaiser: Did a survey -- 1.50 doesn't have Context, 1.51 through 1.57 build cleanly in C++11 and C++14.
[13 Mar 16 15:34] * zao * hkaiser: 1.58 is the first one to fail with C++14 while succeeding in C++11.
[13 Mar 16 15:34] * hkaiser * zao: ok, thanks 
[13 Mar 16 15:34] * zao * hkaiser: Interestingly enough 1.59 succeeds.
[13 Mar 16 15:35] * zao * So 1.58 and 1.60 are unusable. We could recommend 1.59 for OSX for the time being.
[13 Mar 16 15:36] * zao * Making a clean build of 1.59 to verify I haven't handfixed something.

@fmilicchio Please try again with any Boost version except 1.58 or 1.60.

**Comment by fmilicchio** [comment]
I cannot downgrade all installations of boost to 1.59, I will need to wait until a patch solves the issue. @sithhell I confirm the problem is present in the current HEAD of master branch.

**Comment by zao** [comment]
I am skeptical about having a single immutable Boost installation.

C++ is sensitive to ODR violations due to differing build settings. HPX deduces that it can be built with C++14 and does so. Boost defaults with current XCode to C++11, so if you've built your Boost or other libraries with C++11, it would be reckless to combine them without figuring out what varies across language versions.

This is not specific to HPX, it's something you need to engineer into your build steps for any set of language features and defines.

In this case, Boost 1.60's Boost.Context is critically broken in C++14 mode due to the test in `boost/context/detail/config.hpp`. This is a completely self-contained bug in Boost, and cannot really be fixed from the HPX side unless we fork the whole library or distribute patches for building 1.60.

I strongly recommend having purpose-built versions of libraries needed for a particular project. Package managers have yet to figure out a solution to the language/build flavors problem.

**Comment by hkaiser** [comment]
This was fixed by merging #2606
